File: CodeTest/src/BaekJoon/Stack/Q2504/Q2504.java
/**
* Author    : Lee In Bok 
* Date      : 2023.12.29(Fri)
* Spend Time: 47m 48s
* Runtime   : 124 ms
* Memory    : 14280 KB
* Algoritm  : Stack
 */

package BaekJoon.Stack.Q2504;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Q2504 {
  public static void main(String[] args) throws IOException{
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    String s = br.readLine().replace("()", "2").replace("[]", "3");;
    Stack<Character> stack = new Stack<>();

    int mul = 1;
    int ans = 0;

    for(int i = 0; i < s.length(); i++){
      char ch = s.charAt(i);

      if(ch == '('){
        mul *= 2;
      } else if(ch == ')'){
        mul /= 2;
      } else if(ch == '['){
        mul *= 3;
      } else if(ch == ']'){
        mul /= 3;
      }

      if(!stack.isEmpty()){
        if(ch == ')' && stack.peek() == '('){
          stack.pop();
          continue;
        } else if(ch == ']' && stack.peek() == '['){
          stack.pop();
          continue;
        }
      }

      if(48 <= ch && ch <= 57){
        ans += mul * Character.getNumericValue(ch);
      } else {
        stack.add(ch);
      }
    }

    System.out.println(stack.isEmpty() ? ans : 0);
  }
}

File: CodeTest/src/BaekJoon/Stack/Q2504/Q2504.py
"""
# Author    : Lee In Bok 
# Date      : 2023.12.29(Fri)
# Spend Time: 47m 48s
# Runtime   : 44 ms
# Memory    : 31120 KB
# Algoritm  : Stack
"""

import sys
s = sys.stdin.readline().replace('()', '2').replace('[]', '3').strip('\n')

ans = 0
mul = 1
stack = []

for ch in s:
  if ch == '(':
    mul *= 2
  elif ch == ')':
    mul /= 2
  elif ch == '[':
    mul *= 3
  elif ch == ']':
    mul /= 3

  if stack:
    if ch == ')' and stack[-1] == '(':
      stack.pop()
      continue
    elif ch == ']' and stack[-1] == '[':
      stack.pop()
      continue
  
  if not ch.isdigit():
    stack.append(ch)
  else:
    ans += mul * int(ch)
  
if stack:
  print(0)
else:  
  print(int(ans))